cmd/gpud/command/join.go

```diff
@@ -0,0 +1,190 @@
+	provider := cliContext.String("provider")
+	providerGPUShape := cliContext.String("provider-gpu-shape")
+	providerInstanceType := cliContext.String("provider-instance-type")
+	xrayNeeded := cliContext.Bool("xray-needed")
+	totalCPU := cliContext.Int64("total-cpu")
+	extraInfo := cliContext.String("extra-info")
+
+	stateFile, err := config.DefaultStateFile()
+	if err != nil {
+		return fmt.Errorf("failed to get state file: %w", err)
+	}
+	db, err := state.Open(stateFile)
+	if err != nil {
+		return fmt.Errorf("failed to open state file: %w", err)
+	}
+	defer db.Close()
+
+	uid, _, err := state.CreateMachineIDIfNotExist(rootCtx, db)
+	if err != nil {
+		return fmt.Errorf("failed to get machine uid: %w", err)
+	}
+
+	type payload struct {
+		ID                   string `json:"id"`
+		ClusterName          string `json:"cluster_name"`
+		PublicIP             string `json:"public_ip"`
+		Provider             string `json:"provider"`
+		ProviderGPUShape     string `json:"provider_gpu_shape"`
+		ProviderInstanceType string `json:"provider_instance_type"`
+		XrayNeeded           bool   `json:"xray_needed"`
+		TotalCPU             int64  `json:"total_cpu"`
+		ExtraInfo            string `json:"extra_info"`
+	}
+	type RespErr struct {
+		Error  string `json:"error"`
+		Status string `json:"status"`
+	}
+	content := payload{
+		ID:                   uid,
+		ClusterName:          clusterName,
+		PublicIP:             publicIP,
+		Provider:             provider,
+		ProviderGPUShape:     providerGPUShape,
+		ProviderInstanceType: providerInstanceType,
+		XrayNeeded:           xrayNeeded,
+		TotalCPU:             totalCPU,
+		ExtraInfo:            extraInfo,
+	}
+	rawPayload, _ := json.Marshal(&content)
+	response, err := http.Post(fmt.Sprintf("https://%s/api/v1/join", endpoint), "application/json", bytes.NewBuffer(rawPayload))
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+	if response.StatusCode != http.StatusOK {
+		body, err := io.ReadAll(response.Body)
+		if err != nil {
+			return fmt.Errorf("error reading response body: %w", err)
+		}
+		var errorResponse RespErr
+		err = json.Unmarshal(body, &errorResponse)
+		if err != nil {
+			return fmt.Errorf("Error parsing error response: %v\nResponse body: %s", err, body)
+		}
+		return fmt.Errorf("failed to join: %v", errorResponse)
+	}
+	return handleJoinResponse(rootCtx, response.Body)
+}
+
+func handleJoinResponse(ctx context.Context, body io.Reader) error {
+	dir, err := untarFiles("/tmp/", body)
+	if err != nil {
+		return err
+	}
+	scriptPath := filepath.Join(dir, "join.sh")
+	return runCommand(ctx, scriptPath, nil)
+}
+
+func untarFiles(targetDir string, body io.Reader) (string, error) {
+	var dir string
+	gzipReader, err := gzip.NewReader(body)
+	if err != nil {
+		return "", fmt.Errorf("failed to create gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+
+	tarReader := tar.NewReader(gzipReader)
+
+	for {
+		header, err := tarReader.Next()
```

Comment:
need a separate function in a separate package, but for now, we can keep everything here :)